pvlib/ivtools/utility.py

```diff
@@ -0,0 +1,433 @@
+    u1[:, 4] = (
+        a0[:, 0] * a0[:, 1] * a0[:, 2] + a0[:, 0] * a0[:, 1] * a0[:, 3]
+        + a0[:, 0] * a0[:, 2] * a0[:, 3] + a0[:, 1] * a0[:, 2] * a0[:, 3])
+
+    left[:, 0] = (a0[:, 0] - a0[:, 1]) * (a0[:, 0] - a0[:, 2]) * \
+        (a0[:, 0] - a0[:, 3]) * (a0[:, 0] - a0[:, 4])
+    left[:, 1] = (a0[:, 1] - a0[:, 0]) * (a0[:, 1] - a0[:, 2]) * \
+        (a0[:, 1] - a0[:, 3]) * (a0[:, 1] - a0[:, 4])
+    left[:, 2] = (a0[:, 2] - a0[:, 0]) * (a0[:, 2] - a0[:, 1]) * \
+        (a0[:, 2] - a0[:, 3]) * (a0[:, 2] - a0[:, 4])
+    left[:, 3] = (a0[:, 3] - a0[:, 0]) * (a0[:, 3] - a0[:, 1]) * \
+        (a0[:, 3] - a0[:, 2]) * (a0[:, 3] - a0[:, 4])
+    left[:, 4] = (a0[:, 4] - a0[:, 0]) * (a0[:, 4] - a0[:, 1]) * \
+        (a0[:, 4] - a0[:, 2]) * (a0[:, 4] - a0[:, 3])
+
+    df[2:-2] = np.sum(-(u1 / left) * ff, axis=1)
+
+    # second derivative
+    u2[:, 0] = (
+        a0[:, 1] * a0[:, 2] + a0[:, 1] * a0[:, 3] + a0[:, 1] * a0[:, 4]
+        + a0[:, 2] * a0[:, 3] + a0[:, 2] * a0[:, 4] + a0[:, 3] * a0[:, 4])
+    u2[:, 1] = (
+        a0[:, 0] * a0[:, 2] + a0[:, 0] * a0[:, 3] + a0[:, 0] * a0[:, 4]
+        + a0[:, 2] * a0[:, 3] + a0[:, 2] * a0[:, 4] + a0[:, 3] * a0[:, 4])
+    u2[:, 2] = (
+        a0[:, 0] * a0[:, 1] + a0[:, 0] * a0[:, 3] + a0[:, 0] * a0[:, 4]
+        + a0[:, 1] * a0[:, 3] + a0[:, 1] * a0[:, 3] + a0[:, 3] * a0[:, 4])
+    u2[:, 3] = (
+        a0[:, 0] * a0[:, 1] + a0[:, 0] * a0[:, 2] + a0[:, 0] * a0[:, 4]
+        + a0[:, 1] * a0[:, 2] + a0[:, 1] * a0[:, 4] + a0[:, 2] * a0[:, 4])
+    u2[:, 4] = (
+        a0[:, 0] * a0[:, 1] + a0[:, 0] * a0[:, 2] + a0[:, 0] * a0[:, 3]
+        + a0[:, 1] * a0[:, 2] + a0[:, 1] * a0[:, 4] + a0[:, 2] * a0[:, 3])
+
+    df2[2:-2] = 2. * np.sum(u2 * ff, axis=1)
+    return df, df2
+
+
+def rectify_iv_curve(voltage, current, decimals=None):
+    """
+    Sort the IV curve data, remove NaNs and negative
+    values, and combine points with duplicate voltage.
+
+    Parameters
+    ----------
+    voltage : numeric [V]
+    current : numeric [A]
+    decimals : int or None, default None
+        number of decimal places to which voltage is rounded to remove
+        duplicated points. If None, no rounding is done.
+
+    Returns
+    -------
+    voltage : numeric [V]
+    current : numeric [A]
+
+    Raises
+    ------
+    ValueError if voltage and current are different length
+
+    Notes
+    -----
+    ``rectify_IV_curve`` ensures that the IV curve lies in the first quadrant
+    of the (voltage, current) plane. The returned IV curve:
+        * increases in voltage
+        * contains no negative current or voltage values
+        * contains no NaNs
+        * contains no points with duplicate voltage values. Where voltage
+          values are repeated, a single data point is substituted with current
+          equal to the average of current at duplicated voltages.
+    """
+
+    if len(voltage) != len(current):
+        raise ValueError('voltage and current must have the same length')
```

Comment:
generally -1 on this kind of check in pvlib. vstack will throw its own informative enough ValueError if they don't match.

Reply:
Hold over from Matlab